client/src/components/HistoryViewer/HistoryViewer.js

```diff
@@ -54,6 +55,27 @@ class HistoryViewer extends Component {
     }
   }
 
+  /**
+   * Only update the component if there has actually been a change to props or state.
+   * Prevents toast from being triggered twice.
+   */
+  shouldComponentUpdate(nextProps, nextState) {
+    let shouldUpdate = false;
+    // eslint-disable-next-line no-restricted-syntax
+    for (const [key, val] of Object.entries(this.props)) {
+      if (nextProps[key] !== val) {
+        shouldUpdate = true;
+      }
+    }
+    // eslint-disable-next-line no-restricted-syntax
+    for (const [key, val] of Object.entries(this.state)) {
+      if (nextState[key] !== val) {
+        shouldUpdate = true;
+      }
+    }
+    return shouldUpdate;
+  }
```

Comment:
I'm not really pleased about this, but without it the component kept updating even though neither the props or state were being updated, and that resulted in displaying the toast twice.

I'm sure there's a simple react-y way to avoid that (e.g. `useEffect()` if this was a functional component) but I'm not sure how with a class component.

Reply:
(updated since then to add type checks because in some cases one or both weren't an object (undefined I think) which results in an error.